### PR TITLE
Get rid of coin flip strategy in p::s::Tria

### LINKED ISSUE #3078: locally_owned_dofs belong to different block do not equal after DoFRenumbering::subdomain_wise and DoFRenumbering::block_wise
The description may not be very clear, but I will try to explain it.

Since `DoFTools::get_subdomain_association` use a coin-flip strategy to decide the ownership of DoFs on subdomain boundary, the DoFs belong to the save block on the subdomain boundary will be owned by different subdomain. After `subdomain_wise` and `block_wise`, DoFs belong to different blocks have different parallel partitions.

This is typically no harm, but when we want to use block preconditioner, problems may occur.
Say we have assembled a block matrix`[A, B; C, D]` and we want to use `[A+D, 0; 0, A+D]` as the preconditioner matrix. This will fail due to that matrix A and D have different parallel distributions.

Note that `p::d::Triangulation` do not have this problem because each DoF on an interface between processors is assigned to the processor with the smallest processor identifier.

Of course, we can use the same strategy as in `p::d::Triangulation`, but it seems like it will affect the load balance when the number of processes is not very large.

Is there another method to this problem?

The following code shows this problem.

```
#include <deal.II/base/mpi.h>

#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/grid_tools.h>
#include <deal.II/grid/tria.h>

#include <deal.II/dofs/dof_handler.h>
#include <deal.II/dofs/dof_renumbering.h>
#include <deal.II/dofs/dof_tools.h>

#include <deal.II/fe/fe_q.h>
#include <deal.II/fe/fe_system.h>

#include <deal.II/distributed/tria.h>

#include <iostream>

int main(int argc, char **argv) {
  dealii::Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);
  const unsigned int n_mpi_processes =
      dealii::Utilities::MPI::n_mpi_processes(MPI_COMM_WORLD);
  const unsigned int this_mpi_process =
      dealii::Utilities::MPI::this_mpi_process(MPI_COMM_WORLD);

  dealii::FESystem<3> fe(dealii::FE_Q<3>(1), 2);

  {
    dealii::Triangulation<3> tria;
    dealii::GridGenerator::hyper_cube(tria, -1, 1);
    tria.refine_global(3);

    dealii::GridTools::partition_triangulation(n_mpi_processes, tria);

    dealii::DoFHandler<3> dh(tria);
    dh.distribute_dofs(fe);

    dealii::DoFRenumbering::subdomain_wise(dh);
    dealii::DoFRenumbering::block_wise(dh);

    const std::vector<dealii::IndexSet> locally_owned_dofs_per_subdomain =
        dealii::DoFTools::locally_owned_dofs_per_subdomain(dh);

    const dealii::types::global_dof_index dofs_per_block = dh.n_dofs() / 2;
    std::vector<dealii::IndexSet> locally_owned_dofs_per_block(2);
    locally_owned_dofs_per_block[0] =
        locally_owned_dofs_per_subdomain[this_mpi_process].get_view(
            0, dofs_per_block);
    locally_owned_dofs_per_block[1] =
        locally_owned_dofs_per_subdomain[this_mpi_process].get_view(
            dofs_per_block, dh.n_dofs());

    // This assertion will fail.
    // Assert((locally_owned_dofs_per_block[0] ==
    // locally_owned_dofs_per_block[1]),
    //        dealii::ExcMessage("locally owned dofs do not equal!!!"));

    std::cout << "Seq, Block 0, Process " << this_mpi_process << " : ";
    locally_owned_dofs_per_block[0].print(std::cout);
    std::cout << "Seq, Block 1, Process " << this_mpi_process << " : ";
    locally_owned_dofs_per_block[1].print(std::cout);
  }

  {
    dealii::parallel::distributed::Triangulation<3> tria(MPI_COMM_WORLD);
    dealii::GridGenerator::hyper_cube(tria, -1, 1);
    tria.refine_global(3);

    dealii::DoFHandler<3> dh(tria);
    dh.distribute_dofs(fe);

    dealii::DoFRenumbering::block_wise(dh);

    const dealii::types::global_dof_index dofs_per_block = dh.n_dofs() / 2;
    std::vector<dealii::IndexSet> locally_owned_dofs_per_block(2);
    locally_owned_dofs_per_block[0] =
        dh.locally_owned_dofs().get_view(0, dofs_per_block);
    locally_owned_dofs_per_block[1] =
        dh.locally_owned_dofs().get_view(dofs_per_block, dh.n_dofs());

    // This assertion will NOT fail.
    Assert((locally_owned_dofs_per_block[0] == locally_owned_dofs_per_block[1]),
           dealii::ExcMessage("locally owned dofs do not equal!!!"));

    std::cout << "Dist, Block 0, Process " << this_mpi_process << " : ";
    locally_owned_dofs_per_block[0].print(std::cout);
    std::cout << "Dist, Block 1, Process " << this_mpi_process << " : ";
    locally_owned_dofs_per_block[1].print(std::cout);
  }

  return 0;
}
```

### LINKED ISSUE #5595: Non-contiguous DoFs distribution in p::s::Tria
Since #5436 (figured out by doing bisect), I observe inconsistent DoFs distribution with p::s::Tria when calling DoFTools::block_wise. The code below (ran with **mpirun -n 2**) produces the following output:

> p::s::Tria, Block 0, Process 0 : {[0,17], [27,38], [45,56], [63,70]}
> p::s::Tria, Block 1, Process 0 : {[0,74]}
> p::s::Tria, Block 0, Process 1 : {[18,26], [39,44], [57,62], [71,124]}
> p::s::Tria, Block 1, Process 1 : {[75,124]}

> p::d::Tria, Block 0, Process 0 : {[0,74]}
> p::d::Tria, Block 1, Process 0 : {[0,74]}
> p::d::Tria, Block 0, Process 1 : {[75,124]}
> p::d::Tria, Block 1, Process 1 : {[75,124]}

Note that block 0 has weird DoFs distribution, whereas block 1 is ordered properly (compare with p::d::Tria).

### PR DESCRIPTION
**Pull request by agrayver**
Previously, #5436 unified DoFs distribution between p::d::Tria and p::s::Tria upon construction. As #5595 shows, this was not enough.
The reason is that get_subdomain_association (which currently still relies on the coin flip strategy) is called by the DoFTools::locally_owned_dofs_per_subdomain, which in turn, is used within the ParallelShared::renumber_dofs. As a result, any renumbering performed after construction becomes broken. 

Fixes #5595, closes #3078

### CODE REVIEW

**Comment by agrayver** [comment]
> (ii) run the test-suite, most likely some tests should be updated due to this change.

Is this process described somewhere? I have not done this before.

**Comment by davydden** [comment]
> Is this process described somewhere?

yes, see "Information for developers->Testsuite" here https://www.dealii.org/developer/index.html

**Comment by masterleinad** [comment]
If this change happens to pass the testsuite, it would be nice if you would add your test in #5595 to it.

**Comment by agrayver** [comment]
The following tests fail:

> 	842 - mpi/locally_owned_dofs_per_subdomain_02.mpirun=2.debug (Failed)
> 	843 - mpi/locally_owned_dofs_per_subdomain_02.mpirun=2.release (Failed)
> 	6111 - bits/subdomain_ids_02.debug (Failed)
> 	6112 - bits/subdomain_ids_02.release (Failed)
> 	6232 - bits/subdomain_ids_03.debug (Failed)
> 	6233 - bits/subdomain_ids_03.release (Failed)
> 	6548 - bits/subdomain_ids_04.debug (Failed)
> 	6549 - bits/subdomain_ids_04.release (Failed)
> 	3318 - metis/metis_02.debug (Failed)
> 	3319 - metis/metis_02.release (Failed)

How do I update and push the output?

**Comment by agrayver** [comment]
@masterleinad added.

**Comment by masterleinad** [comment]
> How do I update and push the output?

You copy the new output from the `tests` directory in your `build` folder (or wherever you ran the testsuite from) to the `tests` folder in your `source` directory. An example would be to execute
```
$ cp tests/mpi/locally_owned_dofs_per_subdomain_02.debug/mpirun=2/output ../tests/mpi/locally_owned_dofs_per_subdomain_02.mpirun=2.output
```
in your `build` directory assuming the `build` folder is a subfolder of the `source` folder.

**Comment by masterleinad** [comment]
You still need to fix the indentation...

**Comment by agrayver** [comment]
I fixed all tests which failed locally, this should be ready now.

File: doc/news/changes/incompatibilities/20171211AlexanderGrayver [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Changed: Change the logic of DoFTools::get_subdomain_association to have the degrees of freedom along a refinement edge be now all given to the processor with the smallest subdomain_id.
```
Comment (masterleinad) [inline doc/news/changes/incompatibilities/20171211AlexanderGrayver]: Would you break this line at about 80 characters?

File: tests/mpi/blockwise_parallel_01.cc [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        DoFTools::locally_owned_dofs_per_subdomain(dh);
+
+    const types::global_dof_index dofs_per_block = dh.n_dofs() / 2;
+    std::vector<IndexSet> locally_owned_dofs_per_block(2);
+    locally_owned_dofs_per_block[0] =
+        locally_owned_dofs_per_subdomain[this_mpi_process].get_view(0, dofs_per_block);
+    locally_owned_dofs_per_block[1] =
+        locally_owned_dofs_per_subdomain[this_mpi_process].get_view(dofs_per_block, dh.n_dofs());
+
+    // This assertion fails
```
Comment (masterleinad) [inline tests/mpi/blockwise_parallel_01.cc]: This comment can probably go or at least say "failed".

**Comment by agrayver** [comment]
> If you feel like it, it would probably make sense to merge your indentation commit into the one were you create the test.

Done.

**Comment by masterleinad** [comment]
Thanks!

**Pull request merged**